archive/demultiplexer.go

```diff
@@ -354,14 +363,21 @@ func (receiver *RegularCollectionReceiver) Write(buf []byte) (int, error) {
 // cause the RegularCollectionReceiver.Read() to receive EOF
 // Close will get called twice, once in the demultiplexer, and again when the restore goroutine is done with its intent.file
 func (receiver *RegularCollectionReceiver) Close() error {
+	// Close must be idempotent and repeat channel closes panic; only do once.
 	receiver.closeOnce.Do(func() {
-		close(receiver.readLenChan)
-		// make sure that we don't return until any reader has finished
-		<-receiver.readBufChan
+		close(receiver.readBufChan)
 	})
 	return nil
 }
 
+// End signals to any waiting readers that there is nothing more to read, then
+// it waits on the readBufChan, which is closed by the reader-side Close()
+// method.
+func (receiver *RegularCollectionReceiver) End() {
+	close(receiver.readLenChan)
```

Comment:
I don't think `End` gets called more than once, but I'm wondering if I should protect this with a `sync.Once` just in case.

Reply:
What would be the downside of using `sync.Once`?  Right now, I would suggest using it, too. This isn't performance critical code, so it doesn't seem like it could really hurt, to me.

Reply:
I don't think there's any downside.  I'll do it in the next commit.